compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -0,0 +1,141 @@
+use rustc_span::{def_id::DefId, sym, MacroKind};
+
+use crate::{lints::NonLocalDefinitionsDiag, LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `non_local_definitions` lint checks for `impl` blocks and `#[macro_export]`
+    /// macro inside bodies (functions, enum discriminant, ...).
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// trait MyTrait {}
+    /// struct MyStruct;
+    ///
+    /// fn foo() {
+    ///     impl MyTrait for MyStruct {}
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Creating non local definitions go against expectation and can create decrepencies
+    /// in tooling. In should be avoided.
+    pub NON_LOCAL_DEFINITIONS,
+    Warn,
+    "checks for non local definitions"
+}
+
+#[derive(Default)]
+pub struct NonLocalDefinitions {
+    is_in_body: u32,
+}
+
+impl_lint_pass!(NonLocalDefinitions => [NON_LOCAL_DEFINITIONS]);
+
+impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
+    fn check_body(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.is_in_body += 1;
+    }
+
+    fn check_body_post(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.is_in_body -= 1;
+    }
+
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if self.is_in_body > 0 {
+            match item.kind {
+                ItemKind::Impl(impl_) => {
+                    // The RFC states:
+                    //
+                    // > An item nested inside an expression-containing item (through any
+                    // > level of nesting) may not define an impl Trait for Type unless
+                    // > either the **Trait** or the **Type** is also nested inside the
+                    // > same expression-containing item.
+                    //
+                    // To achieve this we get try to get the paths of the _Trait_ and
+                    // _Type_, and we look inside thoses paths to try a find in one
+                    // of them a type whose parent is the same as the impl definition.
+                    //
+                    // If that's the case this means that this impl block decleration
+                    // is using local items and so we don't lint on it.
+
+                    let parent_impl = cx.tcx.parent(item.owner_id.def_id.into());
+
+                    let self_ty_has_local_parent = if let TyKind::Path(QPath::Resolved(
+                        _,
+                        self_ty_path,
+                    )) = impl_.self_ty.kind
+                    {
+                        has_local_parent(self_ty_path, cx, parent_impl)
+                    } else {
+                        false
+                    };
+
+                    let of_trait_has_local_parent = if let Some(of_trait) = impl_.of_trait {
+                        has_local_parent(of_trait.path, cx, parent_impl)
```

Comment:
You should not be looking into generic args of the trait, actually. See this example:
```rust
fn main() {
    dbg!(().as_ref()); // prints `Cat`
}

fn bad() {
    #[derive(Debug)]
    struct Cat;
    impl AsRef<Cat> for () {
        fn as_ref(&self) -> &Cat { &Cat }
    }
}
```
It currently [compiles](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dcbf51109d63ac8f7e2e9cfd7d0bc63e) and the impl inside the `bad` function obviously affects the inference in `main`.

Reply:
I couldn't come up with an example where the same problem happens for the opposite case though (for `Box<Cat>`)...